experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,131 @@
 
 //! Trusted Shuffler implementation.
 
-// TODO(#2614): Implement Trusted Shuffler logic.
+#[cfg(test)]
+mod tests;
+
+use anyhow::{anyhow, Context};
+use futures::future::join_all;
+use rand::{seq::SliceRandom, thread_rng};
+use std::{
+    future::Future,
+    marker::{Send, Sync},
+    mem::replace,
+    ops::DerefMut,
+    sync::{Arc, Mutex},
+};
+use tokio::sync::oneshot;
+
+type Request = Vec<u8>;
+type Response = Vec<u8>;
+
+struct Message {
+    // Arbitrary message data.
+    data: Vec<u8>,
+    // Channel for sending responses back to the client async tasks.
+    response_sender: oneshot::Sender<Response>,
+}
+
+// Trusted Shuffler inmpelentation.
+pub struct TrustedShuffler<F, O>
+where
+    F: Send + Sync + 'static + FnOnce(Request) -> O,
+    O: Send + Sync + Future<Output = Response>,
```

Comment:
I would suggest defining your own trait (with a single function) at some point, instead of this nontrivial bound on a FnOnce. You probably will want to encapsulate the sending logic in a struct anyways, so it would feel natural to impl that trait for it.

Reply:
Async functions in traits are not supported yet: https://doc.rust-lang.org/error-index.html#E0706

Reply:
Yes, I used `async_trait` too in UWABI. Here the [PR](https://github.com/project-oak/oak/pull/2639/files).

Reply:
Done